noir-projects/aztec-nr/aztec/src/utils.nr

```diff
@@ -0,0 +1,80 @@
+use dep::protocol_types::traits::Eq;
+
+mod test;
+
+// Collapses an array of Options with sparse Some values into a BoundedVec, essentially unwrapping the Options and
+// removing the None values. For example, given:
+//   input: [some(3), none(), some(1)]
+// this returns 
+//   collapsed: [3, 1]
+pub fn collapse<T, N>(input: [Option<T>; N]) -> BoundedVec<T, N> where T: Eq {
+    let (collapsed, collapsed_to_input_index_mapping) = collapse_helper(input);
+    collapse_contrain(input, collapsed, collapsed_to_input_index_mapping);
+    collapsed
+}
+
+fn collapse_contrain<T, N>(
+    input: [Option<T>; N],
+    collapsed: BoundedVec<T, N>,
+    collapsed_to_input_index_mapping: BoundedVec<u32, N>
+) where T: Eq {
+    // collapsed should be a BoundedVec with all the non-none elements in input, in the same order. We need to lay down
+    // multiple constraints to guarantee this.
+
+    // First we check that the number of elements is correct
+    let mut count = 0;
+    for i in 0..N {
+        if input[i].is_some() {
+            count += 1;
+        }
+    }
+    assert_eq(count, collapsed.len(), "Wrong collapsed vec length");
+
+    // Then we check that all elements exist in the original array, and are in the same order. To do this we use the
+    // auxiliary collapsed_to_input_index_mapping array, which at index n contains the index in the input array that 
+    // corresponds to the collapsed entry at index n.
+    // Example:
+    //  - input: [some(3), none(), some(1)]
+    //  - collapsed: [3, 1]
+    //  - collapsed_to_input_index_mapping: [0, 2]
+    // These two arrays should therefore have the same length.
+    assert_eq(collapsed.len(), collapsed_to_input_index_mapping.len(), "Collapse hint vec length mismatch");
+
+    // We now look at each collapsed entry and check that there is a valid equal entry in the input array.
+    let mut last_index = Option::none();
+    for i in 0..N {
+        if i < collapsed.len() {
+            let input_index = collapsed_to_input_index_mapping.get_unchecked(i);
+            assert(input_index < N, "Out of bounds index hint");
+
+            assert_eq(collapsed.get_unchecked(i), input[input_index].unwrap(), "Wrong collapsed vec content");
+
+            // By requiring increasing input indices, we both guarantee that we're not looking at the same input
+            // element more than once, and that we're going over them in the original order.
+            if last_index.is_some() {
+                assert(input_index > last_index.unwrap_unchecked(), "Wrong collapsed vec order");
+            }
+            last_index = Option::some(input_index);
+        }
```

Comment:
Think it would handle the case that @sirasistant points to if an else here would simply check for the ones from `collapsed.len()..N`